skfda/preprocessing/dim_reduction/projection/_fpca.py

```diff
@@ -1,33 +1,445 @@
+        if self.n_components > X.n_samples:
+            raise AttributeError("The sample size must be bigger than the "
+                                 "number of components")
+
+        # check that we do not exceed limits for n_components as it should
+        # be smaller than the number of attributes of the funcional data object
+        if self.n_components > X.data_matrix.shape[1]:
+            raise AttributeError("The number of components should be "
+                                 "smaller than the number of discretization "
+                                 "points of the functional data object.")
+
+        # data matrix initialization
+        fd_data = np.squeeze(X.data_matrix)
+
+        # get the number of samples and the number of points of descretization
+        n_samples, n_points_discretization = fd_data.shape
+
+        # if centering is True then subtract the mean function to each function
+        # in FDataBasis
+        if self.centering:
+            meanfd = X.mean()
+            # consider moving these lines to FDataBasis as a centering function
+            # subtract from each row the mean coefficient matrix
+            fd_data -= np.squeeze(meanfd.data_matrix)
+
+        # establish weights for each point of discretization
+        if not self.weights:
+            # sample_points is a list with one array in the 1D case
+            # in trapezoidal rule, suppose \deltax_k = x_k - x_{k-1}, the weight
+            # vector is as follows: [\deltax_1/2, \deltax_1/2 + \deltax_2/2,
+            # \deltax_2/2 + \deltax_3/2, ... , \deltax_n/2]
+            differences = np.diff(X.sample_points[0])
+            self.weights = [sum(differences[i:i + 2]) / 2 for i in
+                            range(len(differences))]
+            self.weights = np.concatenate(([differences[0] / 2], self.weights))
+
+        weights_matrix = np.diag(self.weights)
+
+        # see docstring for more information
+        final_matrix = fd_data @ np.sqrt(weights_matrix) / np.sqrt(n_samples)
+
+        self.pca_ = PCA(n_components=self.n_components)
+        self.pca_.fit(final_matrix)
+        self.components_ = X.copy(data_matrix=self.pca_.components_)
+        self.component_values_ = self.pca_.singular_values_ ** 2
+
+        return self
+
+    def transform(self, X, y=None):
+        """Computes the n_components first principal components score and
+        returns them.
+
+        Args:
+            X (FDataGrid):
+                the functional data object to be analysed
+            y (None, not used):
+                only present because of fit function convention
+
+        Returns:
+            (array_like): the scores of the data with reference to the
+            principal components
+        """
+
+        # in this case its the coefficient matrix multiplied by the principal
+        # components as column vectors
+        return X.copy(data_matrix=np.squeeze(X.data_matrix) @ np.transpose(
```

Comment:
Why do you need `squeeze`? I prefer to know the dimensions of the data.

Reply:
If we pass the data matrix directly, the dimension is NxMx1, as a 3D array. Therefore we have to remove the inner-most dimension. Maybe you have a better method of doing this?

Reply:
If it is necessary to drop dimensions I prefer `reshape`, in order to be sure of which dimensions are kept.

Reply:
Yeah that is better, done